vm/systemSmartContracts/esdt.go

```diff
@@ -2050,6 +2204,214 @@ func (e *esdt) stopNFTCreateForever(args *vmcommon.ContractCallInput) vmcommon.R
+
+	return tokenIdentifier, token, vmcommon.Ok
+}
+
+func (e *esdt) registerDynamic(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	_, _, returnCode := e.createDynamicToken(args)
+	return returnCode
+}
+
+func (e *esdt) registerAndSetAllRolesDynamic(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	tokenIdentifier, token, returnCode := e.createDynamicToken(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	allRoles, err := e.getAllRolesForTokenType(string(token.TokenType))
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	properties, returnCode := e.setRolesForTokenAndAddress(token, args.CallerAddr, allRoles)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	returnCode = e.prepareAndSendRoleChangeData(tokenIdentifier, args.CallerAddr, allRoles, properties)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	err = e.saveToken(tokenIdentifier, token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) checkRolesAreCompatibleToChangeToDynamic(token *ESDTDataV2) error {
+	mapOfRoles := make(map[string]uint32)
+
+	for _, esdtRole := range token.SpecialRoles {
+		for _, role := range esdtRole.Roles {
+			mapOfRoles[string(role)]++
+		}
+	}
+
+	rolesWithHaveToBeSingular := rolesForDynamicWhichHasToBeSingular()
+	for _, role := range rolesWithHaveToBeSingular {
+		if mapOfRoles[role] > 1 {
+			return vm.ErrCannotChangeToDynamic
+		}
+	}
+
+	return nil
+}
+
+func (e *esdt) changeToDynamic(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !e.enableEpochsHandler.DynamicESDTEnabled() {
+		e.eei.AddReturnMessage("invalid method to call")
+		return vmcommon.UserError
+	}
+
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	if bytes.Equal(token.TokenType, []byte(core.FungibleESDT)) {
+		e.eei.AddReturnMessage("cannot change fungible tokens to dynamic")
+		return vmcommon.UserError
+	}
+	if isDynamicTokenType(token.TokenType) {
+		e.eei.AddReturnMessage("tokenID is already dynamic")
+		return vmcommon.UserError
+	}
+
+	err := e.checkRolesAreCompatibleToChangeToDynamic(token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	token.TokenType = append([]byte(dynamic), token.TokenType...)
+
+	err = e.saveToken(args.Arguments[0], token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	logEntry := &vmcommon.LogEntry{
+		Identifier: []byte(args.Function),
+		Address:    args.CallerAddr,
+		Topics:     [][]byte{args.Arguments[0], token.TokenName, token.TickerName, token.TokenType},
+	}
+	e.eei.AddLogEntry(logEntry)
+
+	e.sendTokenTypeToSystemAccounts(args.CallerAddr, args.Arguments[0], token)
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) sendTokenTypeToSystemAccounts(caller []byte, tokenID []byte, token *ESDTDataV2) {
```

Comment:
Let's not forget about unit/integration tests for the new code

Reply:
doing it.